### ISSUE TITLE
Tests having TaskLike objects as their return type throws Exception

### ISSUE THREAD

**Issue by miyaniv** [nunit/nunit#3093]
After upgrading NUnit.3.10.1 to NUnit.3.11.0 tests which returned TaskLike objects have failed.

Message: System.InvalidOperationException: The delegate did not return a Task.

StackTrace:
AsyncToSyncAdapter.Await(Func`1 invoke)
TestMethodCommand.RunTestMethod(TestExecutionContext context)

The tests have passed before the upgrade without fail.

I looked at the code and it throws this exception if returnType FullName is Task's type name.

A TaskLike object is defined by a certain attribute which force the Task interface.
The attribute:
System.Runtime.CompilerServices.AsyncMethodBuilderAttribute

The comment in the code applies this feature is soon to come however I couldn't find an open issue about it.
I would like to know when to expect a resolution.

This issue is important because many of our tests fail due to this behaviour.

**Comment by mikkelbu** [comment]
Hi @miyaniv 
The issue to watch is this one - I guess - nunit/nunit#2286, but I don't know the status of it. Probably @jnm2 will have more information.

**Comment by jnm2** [comment]
Oh, yes, I would like to get this done.

To be technical about it, we ignore the `AsyncMethodBuilderAttribute` attribute (except that we use it to error on async void methods). Is we depended on this attribute, we wouldn't recognize this as a tasklike method:

```cs
[Test]
public Task Test1() => AssertFoo(() => 42);

private async Task AssertFoo(Func<int> foo)
{
    // ...
}
```

Same for ValueTask.

Instead, what we'll do is recognize the C# awaitable pattern as defined in the C# 5.0 spec and unaltered as of C# 7.3: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#awaitable-expressions

> I would like to know when to expect a resolution.
>
> This issue is important because many of our tests fail due to this behaviour.

We're folks like yourself who contribute in our spare time. I can't promise, but I can say that I have prototypes and it makes sense to finish them soon now that we've gotten our second request for this. Thanks for letting us know that this one matters to you!

**Comment by miyaniv** [comment]
Thank you for your quick response

Let me give you a little brief about the Task which may explain why our Tests were fine before and why the workaround is a bit complex for us.

I wrote a Task called ContextTask.
A Regular Task copies its context with **shallow copy** which means not all context is being past to the new Task. 
ContextTask **deep copies** that context.
ContextTask is basically a Task for all that matters. 
- Behind the scenes  he's a normal Task
- GetAwaiter returns Task.GetAwaiter
- And all it's methods and properties are "Inherited" from Task.
- Even the StateMachine is similar

The only difference from a Task is its MethodBuilder.
- Added a method that occurs Before the execution
- Added a method that occurs On Executing Start

Actually the TaskLike is not really a ContextTask it's a TaskWrapper which acts differently in each of our softwares.

Due to this TaskLike having Task's Awaiter we were able to pass tests awaiting TaskWrapper

And About the workaround:
- SetUp initializes context which would get lost if Task was the return type of the Test.
```csharp
invocationResult = invoke.Invoke();
if (invocationResult == null || !IsTaskType(invocationResult.GetType()))
    throw new InvalidOperationException("The delegate did not return a Task."); // General awaitable support coming soon.
```
In some tests we write the the Arrange and Act asynchronously while the Assert
`Assert.That(result, Is.EqualTo(expected)`
Debugging the Test everything works fine even the Assert and outside the scope of Test This exception was thrown
So i went and looked what does IsTaskLike do
```csharp
for (; type != null; type = type.GetTypeInfo().BaseType)
{
    if (type.GetTypeInfo().IsGenericType
        ? type.GetGenericTypeDefinition().FullName == "System.Threading.Tasks.Task`1"
        : type.FullName == "System.Threading.Tasks.Task")
    {
        return true;
    }
}
```
And it seems like it is trying to look for an ancestor which FullName is one of the original Tasks
a TaskLike class doesn't have to derive from Task.
I remember `csharp Task` and `csharp Task<T>` being sealed classes
and now that i see that Task is not a sealed i'll try deriving from it (It might solve it in my case) however classes like ValueTask which does not derive from Task would still fail.

### Short Example
```csharp
[Test]
public async TaskLike DoSomthing__ShouldChangeSomthingElse(){
    var input = expected;
    
    await DoSomthingReturnTaskLike(input);
    
    Assert.That(somthingElse.Value, Is.EqualTo(expected));
}
```

**Comment by jnm2** [comment]
@miyaniv Note that the method is not named `IsTaskLike`; it's named `IsTaskType` and its usage is followed by a comment indicating that it's a temporary situation:

https://github.com/nunit/nunit/blob/bd780c0f32e86bacf165b0b43ec73ba63ff8519b/src/NUnitFramework/framework/Internal/AsyncToSyncAdapter.cs#L78-L79

However, I did not intend there to be a regression from 3.10.1 to 3.11.0. When you mentioned that your tests worked on 3.10.1 and failed when you upgraded to 3.11.0, is that when you had ContextTask derive from System.Threading.Tasks.Task?

I'm not sure I understand why you need ContextTask to be the return type. There should always be a somewhat trivial way to adapt the same code to return `Task`, though there may then be extra boilerplate. If we wanted to dig in and understand your ContextTask, I suspect we would suggest clearer, less "magic" code.
Either way, this will all be fixed soon when NUnit is smart enough to await all C# awaitable types.

This is another technicality, but NUnit soon will support all *awaitable* types which is not the same thing as supporting all `tasklike` types. ‘Awaitable’ refers to *consuming* the asynchronous object (`await` keyword), and ‘tasklike’ refers to *producing* the asynchronous object (`async` keyword). For example, YieldAwaiter is awaitable because it can be used with `await`, but it is not tasklike because you can not create an async method with the return type `YieldAwaitable`.

In theory it's possible to create a tasklike type that is not awaitable, similar to how `async void` is not awaitable, but that would not likely be useful. Therefore, any tasklike type will almost certaianly be awaitable, and any awaitable type will be supported by NUnit soon as a method return type.

**Comment by miyaniv** [comment]
Thanks for prioritizing the issue and listening to what i had to say.

We're using a Context which doesn't flow between threads upon await Task or Task.Run.
The context is only ment to flow upon await ContextTask or ContextTask.Run

In order to have a Task as a return type I'll have to call a SetUpContext method in the begining of every Test.

```csharp
[SetUp]
public void SetUp() => SetUpContext();

public void SetUpContext() => Context.Set(context);

[Test]
public async Task DoSomthing__ShouldChangeSomthingElse(){
    SetUpContext(); // This line should be added to the beginning of every Test
    var input = expected;
    
    await DoSomthingReturnTaskLike(input);  // This code accesses Context
    
    Assert.That(somthingElse.Value, Is.EqualTo(expected));
}
```

> However, I did not intend there to be a regression from 3.10.1 to 3.11.0. When you mentioned that your tests worked on 3.10.1 and failed when you upgraded to 3.11.0, is that when you had ContextTask derive from System.Threading.Tasks.Task?

ContextTask does not derive from Task because there is a ContextTask<> which derive from ContextTask and Can't derive from both ContextTask and Task<> at the same time.

> Either way, this will all be fixed soon when NUnit is smart enough to await all C# awaitable types.

This will surely solve the issue and I admit this sounds a way better solution than what i thought 😄

**Comment by jnm2** [comment]
Fixed in the prerelease feed and included in 3.12 which is just around the corner!
